src/main/java/org/truffleruby/extra/ffi/Pointer.java

```diff
@@ -175,7 +177,10 @@ public synchronized void enableAutorelease(FinalizationService finalizationServi
             return;
         }
 
-        finalizationService.addFinalizer(this, Pointer.class, () -> free());
+        // We must be careful here that the finalizer does not capture the Pointer itself that we'd
+        // like to finalize.
+        long pointer = address;
+        finalizationService.addFinalizer(this, Pointer.class, () -> UNSAFE.freeMemory(pointer));
```

Comment:
Would it be safer to use a `static` method here, passing the address as a local variable, similar to the Ruby idiom?
I am not sure what the semantics of Java lambdas are regarding how much is captured.